### PR TITLE
New: Add pushsafer notification service

### LINKED ISSUE #8329: Add Pushsafer Notification Service
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Is your feature request related to a problem? Please describe

no

### Describe the solution you'd like

please add Pushsafer Notification Service

API description can be found here:
https://www.pushsafer.com/en/pushapi

### Describe alternatives you've considered

please add Pushsafer Notification Service

API description can be found here:
https://www.pushsafer.com/en/pushapi

### Anything else?

https://www.pushsafer.com/en/pushapi

### PR DESCRIPTION
**Pull request by ricci2511**
#### Database Migration
NO

#### Description
Added support for [pushsafer](https://www.pushsafer.com/en/pushapi).

#### Todos
- [ ] [Wiki Updates](https://wiki.servarr.com)

#### Issues Fixed or Closed by this PR

* Closes #8329

### CODE REVIEW

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        [FieldDefinition(1, Label = "Device IDs", HelpText = "Device Group ID or list of Device IDs (leave blank to send to all devices)", Type = FieldType.Tag, Placeholder = "123456789|987654321")]
+        public IEnumerable<string> DeviceIds { get; set; }
+
+        [FieldDefinition(2, Label = "Priority", Type = FieldType.Select, SelectOptions = typeof(PushsaferPriority))]
+        public int Priority { get; set; }
+
+        [FieldDefinition(3, Label = "Retry", Type = FieldType.Textbox, HelpText = "Interval to retry Emergency alerts, minimum 60 seconds")]
+        public int Retry { get; set; }
+
+        [FieldDefinition(4, Label = "Expire", Type = FieldType.Textbox, HelpText = "Maximum time to retry Emergency alerts, maximum 10800 seconds")]
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:42]: Should some of these be advanced settings?

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        public int Expire { get; set; }
+
+        [FieldDefinition(5, Label = "Sound", Type = FieldType.Textbox, HelpText = "Notification sound 0-62 (leave blank to use the default)", HelpLink = "https://www.pushsafer.com/en/pushapi_ext#API-S")]
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:45]: Do we need validation for the range here? Or does it change often?

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        public string Sound { get; set; }
+
+        [FieldDefinition(6, Label = "Vibration", Type = FieldType.Textbox, HelpText = "Vibration pattern 1-3 (leave blank to use the device default)", HelpLink = "https://www.pushsafer.com/en/pushapi_ext#API-V")]
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:48]: Same re. validation

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        public string Vibration { get; set; }
+
+        [FieldDefinition(7, Label = "Icon", Type = FieldType.Textbox, HelpText = "Icon number 1-181 (leave blank to use the default Pushsafer icon)", HelpLink = "https://www.pushsafer.com/en/pushapi_ext#API-I")]
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:51]: Same re. validation

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        public string Icon { get; set; }
+
+        [FieldDefinition(8, Label = "Icon Color", Type = FieldType.Textbox, HelpText = "Icon color in hex format (leave blank to use the default Pushsafer icon color)", HelpLink = "https://www.pushsafer.com/en/pushapi_ext#API-C")]
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:54]: Should we validate this via num characters or something similar?

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    public class PushsaferSettings : IProviderConfig
+    {
+        private static readonly PushsaferSettingsValidator Validator = new ();
+
+        public PushsaferSettings()
+        {
+            Priority = 0;
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:26]: Use the Enum

**Review by Qstick — CHANGES_REQUESTED** [review]
Looks pretty good, just a couple questions/changes

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs (hunk 5 above)
Comment (ricci2511) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:54]: > Should we validate this via num characters or something similar?

Regex?

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs (hunk 1 above)
Comment (ricci2511) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:42]: Makes sense, will do.

Comment (Qstick) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:42]: Makes sense to leave those two as normal in that case, how about the others

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs (hunk 5 above)
Comment (Qstick) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:54]: Maybe?

`RuleFor(c => c.IconColor).NotEmpty().Matches(new Regex("^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$"))`

Comment (ricci2511) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:54]: I'm not an expert with `FluentValidation` but this makes the icon color field mandatory, right? As for the API it's an optional property and for that reason I went with `When(c => !string.IsNullOrWhiteSpace(c.IconColor))` instead.

Comment (Qstick) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:54]: Yep, my bad. But you should use the extension function. `c.IconColor.IsNotNullOrWhitespace()`

Comment (ricci2511) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:54]: For some reason the extension function `IsNotNullOrWhiteSpace` cannot be used, hence `string.IsNullOrWhiteSpace`.

Comment (Qstick) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:59]:
```
using NzbDrone.Common.Extensions;

.When(c => c.IconColor.IsNotNullOrWhiteSpace())
```

File: src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs (hunk 1 above)
Comment (ricci2511) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:48]: Honestly not sure which ones we could consider as "advanced". Maybe device IDs or any of the optional ones like sound, vibration, icon and icon color.

Comment (Qstick) [reply src/NzbDrone.Core/Notifications/Pushsafer/PushsaferSettings.cs:48]: `sound, vibration, icon and icon color`

These were the ones I was thinking

**Review by Qstick — APPROVED** [review]

**Pull request merged**